barter/examples/custom_strategy.rs

```diff
@@ -0,0 +1,414 @@
+};
+use barter_execution::{
+    balance::Balance,
+    client::mock::MockExecutionConfig,
+    order::{ClientOrderId, Order, OrderKind, RequestCancel, RequestOpen, StrategyId, TimeInForce},
+    AccountEvent, AccountEventKind,
+};
+use barter_instrument::{
+    asset::AssetIndex,
+    exchange::{ExchangeId, ExchangeIndex},
+    index::IndexedInstruments,
+    instrument::{
+        kind::InstrumentKind,
+        spec::{
+            InstrumentSpec, InstrumentSpecNotional, InstrumentSpecPrice, InstrumentSpecQuantity,
+            OrderQuantityUnits,
+        },
+        Instrument, InstrumentIndex,
+    },
+    Side, Underlying,
+};
+use barter_integration::channel::{mpsc_unbounded, ChannelTxDroppable, Tx};
+use fnv::FnvHashMap;
+use futures::StreamExt;
+use rust_decimal::Decimal;
+use rust_decimal_macros::dec;
+use tracing::{debug, info, warn};
+
+const EXCHANGE: ExchangeId = ExchangeId::BinanceSpot;
+const RISK_FREE_RETURN: Decimal = dec!(0.05);
+const MOCK_EXCHANGE_ROUND_TRIP_LATENCY_MS: u64 = 100;
+const MOCK_EXCHANGE_FEES_PERCENT: Decimal = dec!(0.05);
+const STARTING_BALANCE_USDT: Balance = Balance {
+    total: dec!(10_000.0),
+    free: dec!(10_000.0),
+};
+const STARTING_BALANCE_BTC: Balance = Balance {
+    total: dec!(0),
+    free: dec!(0),
+};
+const STARTING_BALANCE_ETH: Balance = Balance {
+    total: dec!(1.0),
+    free: dec!(1.0),
+};
+const STARTING_BALANCE_SOL: Balance = Balance {
+    total: dec!(10.0),
+    free: dec!(10.0),
+};
+
+/// Engine state used for running the custom strategy
+type EngineState =
+    state::EngineState<DefaultMarketData, BuyAndHoldStrategyState, DefaultRiskManagerState>;
+
+struct BuyAndHoldStrategy {
+    /// Strategy Id
+    id: StrategyId,
+    /// Instrument traded by this strategy
+    instrument: InstrumentIndex,
+}
+
+#[derive(Debug, Default, Clone)]
+struct BuyAndHoldStrategyState {
+    // Desired quantity we would like to buy
+    desired_quantity_to_buy: Decimal,
+}
+
+impl Processor<&AccountEvent> for BuyAndHoldStrategyState {
+    type Output = ();
+
+    fn process(&mut self, event: &AccountEvent) -> Self::Output {
+        if let AccountEventKind::Trade(trade) = &event.kind {
+            if self.desired_quantity_to_buy.is_zero() {
+                // When this executes it means that we proposed multiple open
+                // orders
+                warn!(?event, "Double spend. Kind of.");
+            }
+
+            self.desired_quantity_to_buy -= trade.quantity;
+        }
+    }
+}
+
+impl Processor<&MarketEvent> for BuyAndHoldStrategyState {
+    type Output = ();
+
+    fn process(&mut self, _event: &MarketEvent) -> Self::Output {
+        // Update strategy state when we receive a specific MarketEvent
+    }
+}
+
+impl AlgoStrategy for BuyAndHoldStrategy {
+    type State = EngineState;
+
+    fn generate_algo_orders(
+        &self,
+        state: &Self::State,
+    ) -> (
+        // TODO: Use default types for orders ExchangeKey and InstrumentKey?
```

Comment:
Yes. 

Strategy currently generates orders with indexes. It has the info to generate with ExchangeId, InstrumentNameExchange but ExecutionTxMap uses Index for faster lookups. 

Perhaps there is some optimisation down the line on this though.